File: src/10-challenge/solution.js
/**
 * @param {string} brackets - The string to check
 * @returns {number} The maximum depth of the magic
 */
export default function fn(brackets) {
	/**@type {string[]} */
	const queue = []
	let deepLevel = -1

	for (const bracket of brackets) {
		if (bracket === '[') {
			queue.push(bracket)
		} else if (queue.length) {
			deepLevel = Math.max(queue.length, deepLevel)
			queue.shift()
		} else return -1
	}

	return queue.length ? -1 : deepLevel
}

File: src/10-challenge/solution.py
from collections import deque


def fn(brackets: str) -> int:
    queue = deque[str]()
    deep_level = -1

    for bracket in brackets:
        if bracket == "[":
            queue.append(bracket)
        elif queue:
            deep_level = max(len(queue), deep_level)
            queue.popleft()
        else:
            return -1

    return -1 if queue else deep_level
